### ISSUE TITLE
Surge, FileSystem, C++-17, etc

### ISSUE THREAD

**Issue by baconpaul** [surge-synthesizer/surge#3589]
Over in https://github.com/surge-synthesizer/shortcircuit3/commit/2f1a3a6ba9ac87dd0c16838f12dc929b63077eaa I moved SC3 to https://github.com/gulrak/filesystem after a couple of conversations with @mvf 

For surge 1.9 I propose the following

1. We use CMAKE_CXX_STANDARD 17 everywhere.
2. This means flooring macOS at 10.12
3. But several of our compilers (macOS < 10.15; gcc < 9) still don't provide std::filesystem with C++-17
4. So we do the same linkage as above optionally on those platforms

Basically: C++ everywhere, still support our user base. 

Doing this is a better idea than what I proposed in #3463 so I will close #3463

**Comment by baconpaul** [comment]
@mvf my intended first commit of surge 1.9 is going to move us to require macOS 10.12 and set our C++ standard to 17.

This means that we will have (basically) c++17-without-filesystem on mac and c++-with-misnamed-filesystem on half our linuxes.

So the only question left on this issue is: do we dump our filesystem and use the gulrak one we are using in SC3 for those platforms? It's easy to do mechanically.

Your thoughts appreciated, especially since it would involve loads of code you have touched.

**Comment by baconpaul** [comment]
Ahh darnit apple silicon C++ 17 will be harder. You need macOS14 to have aligned dealloc in the arm std lib.

Window Oscillator threw an error in the M1 build that 10.12 doesn't have the dealloc. I could fix it by adding an operator new/delete like this but we need to add it to several classes so it won't be the first install

will def do it though.

```
+void* operator new(std::size_t size, std::align_val_t align) {
+    auto ptr = aligned_alloc(static_cast<std::size_t>(align), size);
+
+    if (!ptr)
+        throw std::bad_alloc{};
+    return ptr;
+}
+
+void operator delete(void* ptr, std::size_t size, std::align_val_t align) noexcept {
+    free(ptr);
+}
+
+void operator delete(void* ptr, std::align_val_t align) noexcept {
+    free(ptr);
+}
```

**Comment by mvf** [comment]
> do we dump our filesystem and use the gulrak one we are using in SC3 for those platforms?

Good point, I'll prepare a change that replaces `Surge::filesystem` with gulrak. We'll keep our CMake infrastructure for include and exports hygiene and the ability to build against gulrak anywhere via CMake cache variable.

At this point, the only semi-reason for `Surge::filesystem` would be the explicit `fs::path` ctor that makes Unicode a tad safer on Windows. But it also makes composing paths from ASCII literals annoying, and this will become much more common in 1.9 as we phase out `std::string` in favor of `fs::path` (basically "Approach A" from [this comment](https://github.com/surge-synthesizer/surge/issues/2579#issuecomment-688421813), except that we'll typically use `filebuf` instead of `fstream`). So with gulrak, we'll have to be more careful when converting string variables to paths, but that's all the more reason to finally get rid of these conversions altogether.

**Comment by baconpaul** [comment]
Yup I already did exactly that cmake work (based on your stuff) over in shortcircuit3 repo by the way.

https://github.com/surge-synthesizer/shortcircuit3/commit/2f1a3a6ba9ac87dd0c16838f12dc929b63077eaa

there's the commit if that saves you some time.

I'm also going to pin a 'status' issue up top here for people who aren't in the blow-by-blow on discord. Lots afoot as we end the 18 cycle.

**Comment by mvf** [comment]
Yeah, the status issue is a good idea. Things are easily missable even with Discord.

To integrate GHC cleanly we can't use their `CMakeLists.txt` because it leaks system headers. We even have to patch at configuration time because they force-export their symbols, polluting our library interface. No biggie, but not just a matter of `add_subdirectory`/`target_link_libraries` either.

**Comment by baconpaul** [comment]
right. for short-circuit I just lived with those problems :)

but doing it better would be welcome. short-circuit barely runs

Alas I looked at a C++17 pure upgrade and it causes a problem on macOS - basically we can do macOS 10.12 on x86 with 17 but need 14 on ARM (or we need to define custom new/delete operators for all our aligned classes on mac arm only).

So I think it looks like our next install is "17 / 14 with filesystem if we need it" still to keep our user base going - which is a bummer.
